File: sdks/python-mcp/python/solvapay_mcp/asgi/mcp_oauth_request.py
"""Host-side `mcpOauthRequest`: discovery + OAuth proxy over one JSON contract."""

from __future__ import annotations

from collections.abc import Mapping
from typing import TypedDict, TypeGuard

from solvapay_mcp.server.helpers import _invoke


class McpOauthRequestConfig(TypedDict, total=False):
    publicBaseUrl: str
    productRef: str
    apiBaseUrl: str
    mcpPath: str
    oauthPaths: dict[str, str]


class McpOauthRequestParams(TypedDict):
    method: str
    path: str
    headers: Mapping[str, str]
    body: str
    config: McpOauthRequestConfig


class McpOauthRequestResult(TypedDict):
    status: int
    headers: dict[str, str]
    body: object


def _is_record(value: object) -> TypeGuard[dict[str, object]]:
    return isinstance(value, dict)


def _as_oauth_result(value: object) -> McpOauthRequestResult:
    if not _is_record(value):
        raise RuntimeError("mcpOauthRequest returned a result without status")
    status = value.get("status")
    if not isinstance(status, int):
        raise RuntimeError("mcpOauthRequest returned a result without status")
    headers: dict[str, str] = {}
    raw_headers = value.get("headers")
    if isinstance(raw_headers, dict):
        for key, header in raw_headers.items():
            if isinstance(header, str):
                headers[str(key)] = header
    return {"status": status, "headers": headers, "body": value.get("body")}


async def mcp_oauth_request(
    params: McpOauthRequestParams,
    client: object,
) -> McpOauthRequestResult:
    if client is None:
        raise RuntimeError("oauth_client is required for mcpOauthRequest")
    config = dict(params["config"])
    config.pop("apiBaseUrl", None)
    envelope = await _invoke(
        client,
        "mcp_oauth_request",
        {
            "method": params["method"],
            "path": params["path"],
            "headers": dict(params["headers"]),
            "body": params["body"],
            "config": config,
        },
    )
    return _as_oauth_result(envelope)

File: sdks/typescript/mcp/src/internal/mcp-oauth-request.ts
/**
 * Host-side `mcpOauthRequest`: one JSON contract for discovery + OAuth
 * proxy routes. Always dispatched through the native client.
 */

import type { OAuthBridgePaths } from '@solvapay/mcp-core'

export type McpOauthRequestConfig = {
  publicBaseUrl: string
  productRef: string
  apiBaseUrl: string
  mcpPath?: string
  oauthPaths?: OAuthBridgePaths
}

export type McpOauthRequestParams = {
  method: string
  path: string
  headers: Record<string, string>
  body: string
  config: McpOauthRequestConfig
}

export type McpOauthRequestResult = {
  status: number
  headers: Record<string, string>
  body: unknown
}

export type McpOauthRequestClient = {
  mcpOauthRequest: (params: {
    method: string
    path: string
    headers: Record<string, string>
    body: string
    config: {
      publicBaseUrl: string
      productRef: string
      mcpPath?: string
      oauthPaths?: OAuthBridgePaths
    }
  }) => Promise<unknown>
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object'
}

function asOauthResult(value: unknown): McpOauthRequestResult {
  if (!isRecord(value) || typeof value.status !== 'number') {
    throw new Error('mcpOauthRequest returned a result without status')
  }
  const headers: Record<string, string> = {}
  if (isRecord(value.headers)) {
    for (const [key, header] of Object.entries(value.headers)) {
      if (typeof header === 'string') headers[key] = header
    }
  }
  return { status: value.status, headers, body: value.body }
}

export async function mcpOauthRequest(
  params: McpOauthRequestParams,
  client: McpOauthRequestClient,
): Promise<McpOauthRequestResult> {
  const { apiBaseUrl, ...config } = params.config
  void apiBaseUrl
  return asOauthResult(
    await client.mcpOauthRequest({
      method: params.method,
      path: params.path,
      headers: params.headers,
      body: params.body,
      config,
    }),
  )
}

export function requireOauthClient(
  client: McpOauthRequestClient | null | undefined,
): McpOauthRequestClient {
  if (client == null) {
    throw new Error(
      'oauthClient is required; pass the SolvaPay API client that implements mcpOauthRequest',
    )
  }
  return client
}
